### PR TITLE
Adding Flutter controller for Both Android and iOS, with video feed using WebRTC

### PR DESCRIPTION
**Pull request by sparsh3dwe**

### CODE REVIEW

**Comment by izivkov** [comment]
I'm trying the Flutter version of the controller. Not really familiar with flutter, but will learn how it works.

What was the process to convert from the pure Android controller to using flutter? Is there an automated process?

**Comment by izivkov** [comment]
I'm having some problem building the flutter app. Perhaps I'm missing a step since I'm new to flutter. Here is what I'm doing:

- I installed flutter on my Linux PC as per the flutter site: https://docs.flutter.dev/get-started/install/linux
- From the flutter directory of the project, I entered:
```
flutter pub get
flutter run lib/main.dart
```
- I selected the phone I wanted to run the app.
- I got the following:
```
Launching lib/main.dart on moto g 8 power in debug mode...
lib/screens/controlSelector.dart:6:8: Error: Error when reading 'lib/screens/OnScreenMode.dart': No such file or directory
import 'OnScreenMode.dart';
       ^
...
```
Could you assist?

**Comment by sparsh3dwe** [comment]
> I'm trying the Flutter version of the controller. Not really familiar with flutter, but will learn how it works.
> 
> What was the process to convert from the pure Android controller to using flutter? Is there an automated process?

No, we developed the whole controller altogether into flutter to support iOS and Android devices simultaneously.

**Comment by sparsh3dwe** [comment]
> I'm having some problem building the flutter app. Perhaps I'm missing a step since I'm new to flutter. Here is what I'm doing:
> 
> * I installed flutter on my Linux PC as per the flutter site: https://docs.flutter.dev/get-started/install/linux
> * From the flutter directory of the project, I entered:
> 
> ```
> flutter pub get
> flutter run lib/main.dart
> ```
> 
> * I selected the phone I wanted to run the app.
> * I got the following:
> 
> ```
> Launching lib/main.dart on moto g 8 power in debug mode...
> lib/screens/controlSelector.dart:6:8: Error: Error when reading 'lib/screens/OnScreenMode.dart': No such file or directory
> import 'OnScreenMode.dart';
>        ^
> ...
> ```
> 
> The file `lib/screens/OnScreenMode.dart` exists in the project
> 
> Could you assist?

can you run `flutter doctor` and share output?

**Comment by izivkov** [comment]
> > I'm having some problem building the flutter app. Perhaps I'm missing a step since I'm new to flutter. Here is what I'm doing:
> > 
> > * I installed flutter on my Linux PC as per the flutter site: https://docs.flutter.dev/get-started/install/linux
> > * From the flutter directory of the project, I entered:
> > 
> > ```
> > flutter pub get
> > flutter run lib/main.dart
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * I selected the phone I wanted to run the app.
> > * I got the following:
> > 
> > ```
> > Launching lib/main.dart on moto g 8 power in debug mode...
> > lib/screens/controlSelector.dart:6:8: Error: Error when reading 'lib/screens/OnScreenMode.dart': No such file or directory
> > import 'OnScreenMode.dart';
> >        ^
> > ...
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > The file `lib/screens/OnScreenMode.dart` exists in the project
> > Could you assist?
> 
> can you run `flutter doctor` and share output?

OK, after some tinkering, I was able to resolve all issues when running "flutter doctor":

```
flutter doctor
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.3.10, on Linux Mint 20.3 5.4.0-135-generic, locale en_CA.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
[✓] Chrome - develop for the web
[✓] Linux toolchain - develop for Linux desktop
[✓] Android Studio (version 2021.3)
[✓] VS Code (version 1.74.3)
[✓] Connected device (2 available)
[✓] HTTP Host Availability

• No issues found!
```

Next, I tried running lib/main.dart from the flutter directory. I got this:
```
izivkov@izivkov-ODROID-H2:~/projects/OpenBot/controller/flutter$ flutter run lib/main.dart
Launching lib/main.dart on moto g 8 power in debug mode...
lib/screens/controlSelector.dart:6:8: Error: Error when reading 'lib/screens/OnScreenMode.dart': No such file or directory
import 'OnScreenMode.dart';
       ^
lib/utils/constants.dart:3:8: Error: Error when reading 'lib/screens/Controller.dart': No such file or directory
import '../screens/Controller.dart';
       ^
lib/screens/controlSelector.dart:44:16: Error: The method 'OnScreenMode' isn't defined for the class 'ControlSelectorState'.
 - 'ControlSelectorState' is from 'package:openbot_controller/screens/controlSelector.dart' ('lib/screens/controlSelector.dart').
Try correcting the name to the name of an existing method, or defining a method named 'OnScreenMode'.
        child: OnScreenMode(widget.updateMirrorView, widget.indicatorLeft,
               ^^^^^^^^^^^^
lib/utils/constants.dart:7:17: Error: Undefined name 'utf8encoder'.
    'a-string': utf8encoder.convert('κόσμε'),
                ^^^^^^^^^^^


FAILURE: Build failed with an exception.

* Where:
Script '/home/izivkov/flutter/packages/flutter_tools/gradle/flutter.gradle' line: 1159

* What went wrong:
Execution failed for task ':app:compileFlutterBuildDebug'.
> Process 'command '/home/izivkov/flutter/bin/flutter'' finished with non-zero exit value 1

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.
> Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 41s
Running Gradle task 'assembleDebug'...                             42.9s
```
What am I doing wrong?

**Comment by sparsh3dwe** [comment]
@izivkov, We're using Jetbrains Intellij iDE for development which automatically ignores the case change in case of file imports, i've changes the file import name to onScreenMode.dart, Please retry.
:-)

File: android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java [hunk 1]
```diff
@@ -136,6 +136,7 @@ public void onServiceFound(NsdServiceInfo service) {
         try {
           if (service.getServiceType().equals(SERVICE_TYPE)
               && service.getServiceName().equals(SERVICE_NAME_CONTROLLER)) {
+            System.out.println("found service");
```
Comment (thias15) [inline android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java:139]: Is this just an artefact from debugging or needed for anything?

File: android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java [hunk 2]
```diff
@@ -289,7 +290,7 @@ void runReceiver(Scanner reader) {
       try {
         while (true) {
           String msg = reader.nextLine().trim();
-
+          System.out.println("msg"+msg);
```
Comment (thias15) [inline android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java:293]: Is this just an artefact from debugging or needed for anything?

File: android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java (hunk 1 above)
Comment (thias15) [reply android/app/src/main/java/org/openbot/env/NetworkServiceConnection.java:139]: If we want to keep it, I would prefer to use Timber for consistency.

File: controller/flutter/android/app/build.gradle [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+apply plugin: 'com.android.application'
+apply plugin: 'kotlin-android'
+apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
+
+android {
+    compileSdkVersion flutter.compileSdkVersion
+    ndkVersion flutter.ndkVersion
+
+    compileOptions {
+        sourceCompatibility JavaVersion.VERSION_1_8
```
Comment (thias15) [inline controller/flutter/android/app/build.gradle:33]: How about we upgrade to Java 11?

File: controller/flutter/android/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        jvmTarget = '1.8'
+    }
+
+    sourceSets {
+        main.java.srcDirs += 'src/main/kotlin'
+    }
+
+    defaultConfig {
+        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
+        applicationId "com.itinker.openbot.controller.openbot_controller"
```
Comment (thias15) [inline controller/flutter/android/app/build.gradle:47]: org.openbot.flutter_controller

File: controller/flutter/android/app/src/debug/AndroidManifest.xml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.itinker.openbot.controller.openbot_controller">
```
Comment (thias15) [inline controller/flutter/android/app/src/debug/AndroidManifest.xml:2]: org.openbot.flutter_controller

File: controller/flutter/android/app/src/main/AndroidManifest.xml [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<manifest xmlns:android="http://schemas.android.com/apk/res/android"
+    package="com.itinker.openbot.controller.openbot_controller">
```
Comment (thias15) [inline controller/flutter/android/app/src/main/AndroidManifest.xml:2]: org.openbot.flutter_controller

File: controller/flutter/android/app/src/main/kotlin/com/itinker/openbot/controller/openbot_controller/MainActivity.kt [hunk 7]
```diff
@@ -0,0 +1,6 @@
+package com.itinker.openbot.controller.openbot_controller
```
Comment (thias15) [inline controller/flutter/android/app/src/main/kotlin/com/itinker/openbot/controller/openbot_controller/MainActivity.kt:1]: org.openbot.flutter_controller

File: controller/flutter/ios/Runner.xcodeproj/project.pbxproj [hunk 8]
```diff
@@ -0,0 +1,551 @@
+// !$*UTF8*$!
```
Comment (thias15) [inline controller/flutter/ios/Runner.xcodeproj/project.pbxproj:1]: Isn't this an automatically generated file? If yes, please remove.

File: controller/flutter/lib/utils/phoneSensorToDualDriveConvertor.dart [hunk 9]
```diff
@@ -0,0 +1,39 @@
+import 'package:openbot_controller/utils/forwardSpeed.dart';
+
+class PhoneSensorToDualDriveConvertor {
```
Comment (thias15) [inline controller/flutter/lib/utils/phoneSensorToDualDriveConvertor.dart:3]: convertor --> converter

**Comment by thias15** [comment]
@izivkov did you manage to test all functionalities?

File: controller/flutter/ios/Runner.xcodeproj/project.pbxproj (hunk 8 above)
Comment (sparsh3dwe) [reply controller/flutter/ios/Runner.xcodeproj/project.pbxproj:1]: No, this is the project file. which is needed for the project to build.

**Comment by thias15** [comment]
@izivkov any success yet?

**Comment by izivkov** [comment]
Sorry, I'm out of the country. Have not tried it lately. What problem are
you experiencing?

**Comment by thias15** [comment]
@izivkov no problems, just thought you wanted to test the flutter app.

**Review by thias15 — APPROVED** [review]

**Pull request merged**